go/common/viewingkey/viewing_key.go

```diff
@@ -271,14 +266,71 @@ func getBytesFromTypedData(typedData apitypes.TypedData) ([]byte, error) {
 	return rawData, nil
 }
 
-// GenerateAuthenticationEIP712RawDataOptions generates all the options or raw data messages (bytes)
-// for an EIP-712 message used to authenticate an address with user
-// (currently only one option is supported, but function leaves room for future expansion of options)
-func GenerateAuthenticationEIP712RawDataOptions(userID string, chainID int64) ([][]byte, error) {
-	if len(userID) != UserIDHexLength {
-		return nil, fmt.Errorf("userID hex length must be %d, received %d", UserIDHexLength, len(userID))
+// CalculateUserIDHex CalculateUserID calculates userID from a public key
+// (we truncate it, because we want it to have length 20) and encode to hex strings
+func CalculateUserIDHex(publicKeyBytes []byte) string {
+	return hex.EncodeToString(CalculateUserID(publicKeyBytes))
+}
+
+// CalculateUserID calculates userID from a public key (we truncate it, because we want it to have length 20)
+func CalculateUserID(publicKeyBytes []byte) []byte {
+	return crypto.Keccak256Hash(publicKeyBytes).Bytes()[:20]
+}
+
+// CheckSignatureAndReturnAccountAddress checks if the signature is valid for hash of the message and checks if
+// signer is an address provided to the function.
+// It returns an address if the signature is valid and nil otherwise
+func CheckSignatureAndReturnAccountAddress(hashBytes []byte, signature []byte) (*gethcommon.Address, error) {
+	pubKeyBytes, err := crypto.Ecrecover(hashBytes, signature)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKey, err := crypto.UnmarshalPubkey(pubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	r := new(big.Int).SetBytes(signature[:32])
+	s := new(big.Int).SetBytes(signature[32:64])
+
+	// Verify the signature and return the result (all the checks above passed)
+	isSigValid := ecdsa.Verify(pubKey, hashBytes, r, s)
+	if isSigValid {
+		address := crypto.PubkeyToAddress(*pubKey)
+		return &address, nil
 	}
-	encryptionToken := "0x" + userID
+	return nil, fmt.Errorf("invalid signature")
+}
+
+type MessageGenerator interface {
```

Comment:
let's move these types to a separate file in the same package.
The viewing_key one is getting a bit large

Reply:
While we're at it, let's move the signing checking in a separate file as well

Reply:
Also, the "GenerateViewingKeyForWallet" logic with the 2 utility methods it uses

Reply:
Good point, added 2 new files:
- viewing_key_messages.go // for all the constants and functions related to generating messages
- viewing_key_signature.go // for all the code related to checking signatures 

And it is much clearer now and easier to see things.